lib/services/local/headlessauthn_watcher.go

```diff
@@ -222,63 +222,87 @@ func (h *HeadlessAuthenticationWatcher) notify(headlessAuthns ...*types.Headless
-	checkBackend := func() (*types.HeadlessAuthentication, bool, error) {
-		headlessAuthn, err := h.identityService.GetHeadlessAuthentication(ctx, name)
-		if err != nil {
-			return nil, false, trace.Wrap(err)
-		}
-
-		ok, err := cond(headlessAuthn)
-		if err != nil {
-			return nil, false, trace.Wrap(err)
+func (h *HeadlessAuthenticationWatcher) Wait(ctx context.Context, name string, cond func(*types.HeadlessAuthentication) (ok bool, err error)) (*types.HeadlessAuthentication, error) {
+	conditionMet := make(chan *types.HeadlessAuthentication, 2)
+	conditionErr := make(chan error, 2)
+	checkCondition := func(ha *types.HeadlessAuthentication) {
+		if ok, err := cond(ha); err != nil {
+			conditionErr <- trace.Wrap(err)
+		} else if ok {
+			conditionMet <- ha
 		}
-
-		return headlessAuthn, ok, nil
-	}
-
-	// Before the waiter is allocated, check if there is an existing entry in the backend.
-	headlessAuthn, ok, err := checkBackend()
-	if err != nil && !trace.IsNotFound(err) {
-		return nil, trace.Wrap(err)
-	} else if ok {
-		return headlessAuthn, nil
 	}
 
-	waiter, err := h.assignWaiter(ctx, name)
+	waiter, err := h.assignWaiter(name)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	defer h.unassignWaiter(waiter)
 
-	for {
+	ctx, cancel := context.WithCancel(ctx)
+	var wg sync.WaitGroup
+	defer func() {
+		cancel()
+		wg.Wait()
+		h.unassignWaiter(waiter)
+	}()
+
+	// Consume the main channel.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case ha := <-waiter.ch:
+				checkCondition(ha)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// Consume the stale channel.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
 		select {
 		case <-waiter.stale:
-			// If the waiter is a slow consumer it may be marked as stale, in which
-			// case it should check the backend for the latest resource version.
-			h.unmarkStale(waiter)
-
-			headlessAuthn, ok, err := checkBackend()
+			ha, err := h.identityService.GetHeadlessAuthentication(ctx, name)
 			if err != nil {
-				return nil, trace.Wrap(err)
-			} else if ok {
-				return headlessAuthn, nil
-			}
-		case headlessAuthn := <-waiter.ch:
-			if ok, err := cond(headlessAuthn); err != nil {
-				return nil, trace.Wrap(err)
-			} else if ok {
-				return headlessAuthn, nil
+				conditionErr <- trace.Wrap(err)
+			} else {
+				checkCondition(ha)
 			}
 		case <-ctx.Done():
-			return nil, ctx.Err()
-		case <-h.closed:
-			return nil, watcherClosedErr
+			return
+		}
+	}()
+
+	// With the waiter allocated and the main channel serviced, check the backend for an existing entry.
+	go func() {
+		ha, err := h.identityService.GetHeadlessAuthentication(ctx, name)
+		if trace.IsNotFound(err) {
+			// Ignore not found errors in the initial stale check.
+			return
+		} else if err != nil {
+			conditionErr <- trace.Wrap(err)
+		} else {
+			checkCondition(ha)
```

Comment:
This makes 3 possible writes to each channel, right? Should we increase the buffer to 3?